Create Dangerzone environments and run QA tests on them

apyrgio opened the pull request:
This PR introduces two scripts that can help with Dangerzone development and releases. Those are:

* `dev_scripts/env.py`: This script creates Dangerzone environments on various Linux platforms. These environments can be used to test Dangerzone and build artifacts.
* `dev_scripts/qa.py`: This script runs QA tests on the environments that the previous script creates. These QA tests are taken from our release instructions, but they are not fully automated. They act as guard rails for the developer who wants to ensure that Dangerzone works for a platform.

Closes #286
Closes #287

deeplow commented:
I had some specific comments about the windows adaptation of the `build-image.py` but I think I have a better suggestion that works on all OSes - `install/build-image.py`.

```python
from dangerzone.container import get_runtime_name
import subprocess

runtime = get_runtime_name()

def run(cmd):
    subprocess.run(cmd, shell=True)

print("Building container image")
run(f'{runtime} build container --platform linux/amd64 --tag dangerzone.rocks/dangerzone')

print("Saving and compressing container image")
run(f'{runtime} save dangerzone.rocks/dangerzone | gzip > share/container.tar.gz')

print("Looking up the image id")
run(f'{runtime} image ls dangerzone.rocks/dangerzone | grep "dangerzone.rocks/dangerzone" | tr -s \' \' | cut -d\' \' -f3 > share/image-id.txt')
```

This way we would de-duplicate some of the code and remove the added `gzip` complexity in python code. I tested on Linux and this python script is 5 times faster than the proposed one.

What do you think about using this one? I can also add this as a separate PR as it's not core this this change.

deeplow reviewed (changes requested):
I tested this solely on a Qubes system. I haven't tested yet on windows.

### Feedback:
  - [ ] **duplication:** see [this discussion](https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1054373702).
  - [ ] **there could be a way to move back to a previous step. From playing around it appears to be easy to mistakenly move on to the next step.
  - [ ] have you considered mounting in the containers `~/.bashrc`? It oculd be useful to have the history of commands so we don't have to repeatedly type them.
  - [ ] The build `.deb` step was much slower for me. It took 27mins. Is this slow performance to be expected?
  - [ ] In the container there is no PDF viewer, which leads to
    ![no-pdf-viewer](https://user-images.githubusercontent.com/47065258/209135525-c9663944-58e1-48aa-bef8-b1c554b2b824.png)
  - when an 'auto' is running and you press any key when the command is running, those keys will get pressed after the command runs. So if you press <kbd>enter</kbd> while waiting to see if something happens, then it presses enter on the next commands -- but because enter skips the task, you advance to the next ones. I don't know how to solve this problem easily, but changing the auto from `enter` to something else less clickable can help.


### Some situations where 'auto' failed:
- I've had many 'auto' commands related to lack of space. But that's to be expected. After increasing the size the went away. Because this may be so common, should we add a check for them?

-  <details><summary>Failure in auto step "Create Dangerzone QA environment" (fedora 36)</summary>

I was unable to find why this was failed, but after repeating the step, it worked.

```
[...]
  xcb-util-renderutil-0.3.9-20.fc36.x86_64                                      
  xcb-util-wm-0.4.1-22.fc36.x86_64                                              
  xkeyboard-config-2.35.1-1.fc36.noarch                                         
  xml-common-0.6.3-58.fc36.noarch                                               
  xz-5.2.5-9.fc36.x86_64                                                        
  yajl-2.1.0-18.fc36.x86_64                                                     

Complete!
--> de82eca9bdf
STEP 5/10: RUN rm /tmp/dangerzone-0.4.0-1.noarch.rpm
error running container: did not get container start message from parent: EOF
Error: building at STEP "RUN rm /tmp/dangerzone-0.4.0-1.noarch.rpm": slirp4netns failed
Traceback (most recent call last):
  File "/home/user/dangerzone/dev_scripts/env.py", line 509, in <module>
    sys.exit(main())
  File "/home/user/dangerzone/dev_scripts/env.py", line 505, in main
    return args.func(args)
  File "/home/user/dangerzone/dev_scripts/env.py", line 409, in env_build
    env.build(show_dockerfile=args.show_dockerfile)
  File "/home/user/dangerzone/dev_scripts/env.py", line 391, in build
    self.runtime_run("build", "-t", image, build_dir)
  File "/home/user/dangerzone/dev_scripts/env.py", line 235, in runtime_run
    subprocess.run(self.runtime_cmd + list(args), check=True)
  File "/usr/lib64/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['podman', 'build', '-t', 'dangerzone.rocks/fedora:36', PosixPath('/home/user/dangerzone/dev_scripts/envs/fedora/36/build')]' returned non-zero exit status 125.
Traceback (most recent call last):
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 470, in <module>
    sys.exit(main())
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 464, in main
    platform.start()
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 294, in start
    self.build_qa_image()
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 91, in inner
    out = func(self, *args, **kwargs)
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 280, in build_qa_image
    self.run(
  File "/home/user/dangerzone/./dev_scripts/qa.py", line 116, in run
    subprocess.run(args, check=True)
  File "/usr/lib64/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('/home/user/dangerzone/dev_scripts/env.py', '--distro', 'fedora', '--version', '36', 'build')' returned non-zero exit status 1.
```

</details>

<details>

**Other suggestions:**

- **copy `test_docs` to the container's home folder**
  When testing I found myself wasting quite a bit of time with selecting the documents. If every time the container I had already the test documents in the home directory, I could select them more easily. Additionally, it would be nice if we had a fresh copy of the `test_docs` directory, since every time convert them, they are moved to the `unsafe` directory.

- **Report generation**
  It would be nice at the end if the script generated some markdown that could be pasted on the QA issue for the release about what was tested. See my comment about `prompt()` on collecting notes about scenarios and then reporting them at the end.
  
  
**Next steps (possibly future PRs):**
  - Replace in CI `build-*` with building with your setup and deprecate packagecloud

apyrgio commented:
> I had some specific comments about the windows adaptation of the `install/windows/build-image.py` but I think I have a better suggestion that works on all OSes - `install/build-image.py`:
> 
> 8< ... snip ... >8

Just to make sure I understand, do you propose this as a drop-in script for all OSes? Because the `gzip` executable will not be available on Windows

> This way we would de-duplicate some of the code and remove the added `gzip` complexity in python code. I tested on Linux and this python script is 3 times faster than the proposed one.

That's weird. The script you suggest and the existing Linux one are very similar, so where do you attribute this boost?

> What do you think about using this one? I can also add this as a separate PR as it's not core this this change.

In general, having a script for all OSes would be cool, so I'd be down with that, as a separate PR of course. We could even branch this discussion to a new issue.

apyrgio commented on .circleci/config.yml:
```diff
@@ -80,15 +80,15 @@ jobs:
             apt-get install -y make python3 python3-pip python3-venv
-            python3 -m venv .venv
-            source .venv/bin/activate
+            python3 -m venv ~/.venv
+            source ~/.venv/bin/activate
```
Yeap, I can remove it on rebase.

apyrgio commented on .circleci/config.yml:
```diff
@@ -80,15 +80,15 @@ jobs:
           command: |
-            source .venv/bin/activate
-            make lint
+            source ~/.venv/bin/activate
+            poetry run make lint
```
(same as above)

deeplow commented:
> > I had some specific comments about the windows adaptation of the `install/windows/build-image.py` but I think I have a better suggestion that works on all OSes - `install/build-image.py`:
> > 8< ... snip ... >8
> 
> Just to make sure I understand, do you propose this as a drop-in script for all OSes? Because the `gzip` executable will not be available on Windows

Ah. Damn. I forgot about that. That's probably the reason why the windows one was in python. I don't know where I concluded that it worked on all systems, then.

Let's do this in another PR as you suggest.


> > This way we would de-duplicate some of the code and remove the added `gzip` complexity in python code. I tested on Linux and this python script is 3 times faster than the proposed one.
> 
> That's weird. The script you suggest and the existing Linux one are very similar, so where do you attribute this boost?


Chunk sizes, perhaps?

apyrgio commented on install/windows/build-image.py:
```diff
@@ -18,28 +18,26 @@ def main():
+    cmd = subprocess.Popen(
         [
             "docker",
             "save",
             "dangerzone.rocks/dangerzone",
-            "-o",
-            "share/container.tar",
-        ]
+        ],
+        stdout=subprocess.PIPE,
```
Using `subprocess.run()`, we would wait for the command to complete. There are two problems with that:

1. If the command writes to a pipe that no one reads from, it will fill after 65KiB of data, and the command will block.
2. Even if the pipe didn't fill, we have more performance gains if we run `docker save` on parallel with `gzip`.

deeplow replied:
Interesting! Good to know :)

apyrgio commented on install/windows/build-image.py:
```diff
@@ -18,28 +18,26 @@ def main():
+            else:
+                break
+
+    cmd.wait(5)
```
`Popen.communicate()` serves a different purpose, but I'll try to demystify this bit:

* `Popen.communicate()` basically writes to the process' **stdin** once, consumes the process' **stdout** / **stderr**, and then returns it to the program once it completes.
* In this snippet, we need to constantly read from the process's **stdout**, gzip it, and write to a file, which is different.

So, we can't use `Popen.communicate()` here, and we have to do pipe mangling on our own. Now, why do the Python docs suggest using `Popen.communicate()` in the first place? Because the use case of "write to stdin, read from stdout/stderr" is a very common need, that is surprisingly easy to get it wrong (see https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1060046132 for more details).

deeplow replied:
It seems I have to learn more about pipes in python! And thanks for the clarification!

I wish there was a construction like `with (result.pipe) as stdout:`... or something more pythonic.

apyrgio commented on dev_scripts/README.md:
```diff
@@ -0,0 +1,153 @@
+   create new namespaces
+   (https://docs.docker.com/engine/security/seccomp/#significant-syscalls-blocked-by-the-default-profile). For testing/development purposes, we can get around this limitation
+   by disabling the seccomp filters for the external container with
+   `--security-opt seccomp=uncofined`. This has the same effect as developing
```
Merci!

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+        if gui:
+            # Detect X11 display server connection settings.
+            env_display = os.environ.get("DISPLAY")
+            env_xauthority = os.environ.get("XAUTHORITY")
```
Wait, that's interesting. We need a different fallback there. If the environment does not have an Xauth token (i.e., `XAUTHORITY` envvar), then most probably we can communicate with the display server without a token. So, can you test the following?

```diff
             run_cmd += [
                 "-e",
                 f"DISPLAY={env_display}",
-                "-e",
-                f"XAUTHORITY={env_xauthority}",
                 "-v",
                 "/tmp/.X11-unix:/tmp/.X11-unix:ro",
-                "-v",
-                f"{env_xauthority}:{env_xauthority}:ro",
             ]
 
+            if env_xauthority:
+                run_cmd += [
+                    "-e",
+                    f"XAUTHORITY={env_xauthority}",
+                    "-v",
+                    f"{env_xauthority}:{env_xauthority}:ro",
+                ]
+
```

This should have less side effects than creating a dummy `.Xauthority` file.

deeplow replied:
You're right! This suggestion works.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+
+    def container_run(self, *args):
+        """Run a command inside a Dangerzone environment."""
+        cmd_str = " ".join(args)
```
Ouch, this is a left-over from a refactor. I'll remove it.

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+        """Run a command in a Dangerzone environment."""
+        # FIXME: Allow wiping the state of the distro before running the environment, to
+        # ensure reproducibility.
+        if not cmd:
+            print("Please provide a command for the environment")
+            sys.exit(1)
```
Well, `argparse.REMAINDER` is lenient, and may return an empty list if the user has not specified something. So, this argument is not really mandatory. In any case, sure, I can move it out.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+                if self.try_auto:
+                    try_auto = True
+                else:
+                    prompt = "Type 'auto' to try automatically, or ENTER to skip: "
```
Two comments here:

1. "Manual" and "Skip" are basically the same thing. In both cases, the program goes to the next step, and we assume that the user has done "something". So, I'd suggest keeping one of these two consistently, preferably "skip".
2. I was hoping that `--try-auto` / `--skip-manual` would fit the bill here, with regards to non-interactivity. If the user is doing things interactively though, maybe it makes sense to not let stray `ENTER` keys perform an action they didn't intend to. May I suggest the following instead?
    * Consume the process' `stdin` before giving a prompt to the user. This way, key presses that were registered in previous steps won't affect future ones.
    * Ask the user to press `a` for "auto" or `s` for "skip".


What do you think?

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+    """Class for the Windows QA tasks."""
+
+    REF_BUILD = (
+        "https://github.com/freedomofpress/dangerzone/blob/main/BUILD.md#windows"
```
First of all, I tried looking for a Python project that does something like the above, but I didn't find anything. Either my Google-fu have failed me, or no (popular) project exists that takes a stab at the problem of code deviation from a source of truth.

In any case, your solution is very cool, and we should definitely adopt it. I see some minor issues though, that I'd like to discuss:

Relying on the Git revision is problematic, because you can't reference a Git hash in the same Git commit. You have to do so in a subsequent commit. Even then, it's not easy to do so from a PR, because the hash may change, e.g. due to a rebase, fixups, etc.

I think we need to use SHA-256 instead for these blocks. They are a bit more difficult to compute (e.g., run `sed -n 3,47p BUILD.md | sha256sum` for the Debian/Ubuntu section on `BUILD.md`), but they don't have these drawbacks.

deeplow replied:
After a sync discussion, we'll do the following:

1. Move check for diffs on instance method instead of main on the Reference class.
2. Specify the whole section in the Reference (this will be verbose)
3. Introduce qa.py --check-references
4. Run qa.py --check-references on our lint checks.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+        description, and asks the user if they pass.
+        """
+        skip = skip or []
+        REF_QA = "https://github.com/freedomofpress/dangerzone/blob/49b7736cb49d8b03d44c66a3903ba4b9ba1f9171/RELEASE.md#scenarios"
```
I agree that the scenarios shouldn't be out of sync as well, but I'd prefer if users could see them, instead of finding them from a dev script. They are useful as they can see at a glance if a change they want to make must be tested also manually, especially if it's a GUI one.

In any case, we can cover these deviations with the solution you provided above.

deeplow replied:
Yes, let's do that.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+    ch.setLevel(lvl)
+
+    # create formatter
+    formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
```
I didn't want to introduce colors right now, because we need special logic when piping the output of this script to a file, stuff that `git log` for instance does automatically. But I'd definitely want to introduce them at some point for visibility.

As for timestamps, I wanted to include them for two reasons:

1. It's an easy way to see how long each section has taken. Some steps (like building an image) take some time. A slow or fast execution may be a sign of something buggy.
2. It makes stdout / stderr from internal processes stand out, since they typically don't have timestamps.

Does this make sense?

deeplow replied:
Sure, no need to add them now. And we can keep the time, but maybe a shorter version so it's not a verbose?

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+
+    @QABase.auto("Build Dangerzone container image", ref=REF_BUILD)
+    def build_image(self):
+        self.run("python", r".\install\windows\build-image.py")
```
No, it's a [raw Python string](https://note.nkmk.me/en/python-raw-string-escape/). They are typically used for regexes, but you can use them whenever you want to treat backslashes (`\`) as literal characters.

deeplow replied:
Duh... (facepalm) I always assumed they were for regexes because of the `r`. But `r` for raw makes much more sense.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+            .strip("\n")
+        )
+
+    def auto(*msgs, ref=None):
```
I had done this in one of the revisions of the script, but I reverted because it seemed like an overkill. You don't gain much because all it does is print the title of the step through the decorator, which you could do already. But now that we are discussing about anchors in the docs, I'll add it.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+]
+
+
+class Platform(enum.Enum):
+    """List of available platforms for QA."""
+
+    WINDOWS = "windows"
+    UBUNTU_22_04 = "ubuntu-22.04"
+    UBUNTU_22_10 = "ubuntu-22.10"
+    FEDORA_36 = "fedora-36"
+    FEDORA_37 = "fedora-37"
```
Sure, makes sense. I can add something similar using the `inspect` module, which will simplify the `get_platforms_cls_map()`.

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+"""
+
+# FIXME: Install Poetry on Fedora via package manager.
+DOCKERFILE_BUILD_DEV_FEDORA_DEPS = r"""
```
If you see in the Dockerfile, there's a backslash (`\`) that we need to preserve. We do so with raw Python strings (see also https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1060788979)

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+        self.distro = distro
+        self.version = version
+
+        # Try to autodetect the runtime, if the user has not provided it.
```
I don't think we can do that, because the developer can just as well use Docker for the **environment** where they will run Dangerzone in. That environment will of course end up having Podman installed there, but that's different. So, what I'm doing here ("use Podman, else Docker, else fail") does not align with Dangerzone's needs.

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+"""
+
+DOCKERFILE_BUILD_DEV_DEBIAN_DEPS = rf"""
+RUN apt-get update && apt-get install -y \
+    podman dh-python python3 python3-stdeb python3-pyside2.qtcore \
+    python3-pyside2.qtgui python3-pyside2.qtwidgets python3-appdirs \
+    python3-click python3-xdg python3-colorama
+
+RUN apt-get update && apt-get install -y make
+RUN apt-get update && \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    python3-dev python3-venv python3-pip
```
Hm, I'm not sold on the above suggestions because I think (1) breaks caching, which we get for free with `docker/podman build` and (2) would make `env.py` more complex.

Yet I totally feel the pain. We need to do something about it, but I hoped we could punt it for the future. The reason is that the problem seems to me much bigger than the drift between `env.py` / `BUILD.md`. We already have lots of places where we define dependencies. Here's a (hopefully) exhaustive list:

* `.circleci/config.yml`
* `BUILD.md`
* `dev_scripts/env.py`
* `install/linux/build-rpm.py`
* `pyproject.toml`
* `stdeb.cfg`

I think we need to approach this a bit differently:

1. What is the role of these dependencies? How do they differ from the ones already specified in `pyproject.toml`?

    My guess is that they are the system equivalent dependencies of the Python ones, and they are there prior to Poetry's introduction (see this [past README.md version](https://github.com/freedomofpress/dangerzone/blob/8a84c115a1af78bcee62bfc8d255a7ff3e99efc7/README.md)).

2. Can we ditch them and just ask the user to do `poetry install ...`?

     I think yes. The main problem is that the development environment will not be on par with the user's environment, but that's not a problem. Part of our QA steps is to check the **installed* Dangerzone application in a clean environment (which `env.py` already provides), so we will catch any errors there.

3. What about `stdeb.cfg` and `install/linux/build-rpm.py`, which refer to the system dependencies?

     Both these files are used by Python code. We can extend this Python code to check that the source of truth (`pyproject.toml`) has no diverged since writing the list of system dependencies. Similar to what you did with the reference trick.

What do you think?

deeplow replied:
After a synchronous discussion, these were our conclusions:

1. Remove deps from .circleci/config.yml / BUILD.md / env.py
2. Move any deps that are not listed on pyproject toml (stdeb?) to pyproject.toml, preferably under a different section (e.g., tools.poetry.group.package).
3. Test these changes.
4. Create an issue for removing the system deps from stdeb.cfg and build-rpm.py (e.g, dh-python could read them from pyproject.toml) (optional)

deeplow commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+    parser_run.add_argument(
+        "--user",
+        "-u",
+        default=DEFAULT_USER,
```
Never mind. I think the Dockerfile already adds [this user](https://github.com/freedomofpress/dangerzone/pull/289/files/87dd94338b3db3f92e2ef67a9f93f8cfe1015729#diff-c265036fda71ec014271062a6409e5b0f75fcf79d50c190d1ce521de651c2668R90). I thought I had removed this comment from the review but apparently I had not.

apyrgio replied:
The `-u` argument is for the user in the Dangerzone environment, and it's not affected by the dev's environment. In the Dangerzone environment, there are actually two users; the `user` one we create when building it (see the Dockerfile), and `root`. I can go with a `--root` boolean argument, if it makes more sense to you.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+        if ref:
+            logger.info("For reference, see: %s" % ref)
+
+    def prompt(self, *msgs, ref=None, prompt=None):
```
Interesting categorization, I agree with that. As for the two suggestions, I think that asking the user for an explanation would probably not amount to much, and will add extra complexity to the code. As for changing this to a `[Y/n]` prompt, I'd agree, since it is aligned with [this discussion](https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1049674100).

My concern is that us devs may have partially run QA on a platform, and may want a quick way to reach the point where we left off, so the `[Y/n]` prompts may prove frustrating. We'll see...

apyrgio commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,503 @@
+DEFAULT_DEV = False
+DEFAULT_SHOW_DOCKERFILE = False
+
+# The Linux distributions that we currently support.
+# FIXME: Add a version mapping to avoid mistakes.
+# FIXME: Maybe create an enum for these values.
+DISTROS = ["debian", "fedora", "ubuntu"]
```
Yeap, I replied to that. I suggest we keep the discussion going on there, and resolve this thread.

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,448 @@
+                # user has run it manually.
+                if try_auto:
+                    logger.info("Trying automatically...")
+                    out = func(self, *args, **kwargs)
```
I used to have this code, but it was it felt a bit tricky to get it right:

* The traceback would appear before a prompt.
* Sometimes you wanted to run a command, meaning that you needed to start a new shell.
* It would have to fail if `--try-auto` was used.

So, I figured that we can get this right in a subsequent iteration, as it's not critical. Are you ok with that?

deeplow replied:
Sure!

apyrgio commented:
Replying to the rest of the feedback (https://github.com/freedomofpress/dangerzone/pull/289#pullrequestreview-1217986488):

> I tested this solely on a Qubes system. I haven't tested yet on windows.
> ### Feedback:
> 
>     * [ ]  **duplication:** see [this discussion](https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1054373702).

Yeap, juicy subject, let's continue the discussion there.
 
>     * [ ]  **there could be a way to move back to a previous step. From playing around it appears to be easy to mistakenly move on to the next step.

We _can_ do it, but is it worth doing it? I think it's one of the things we should do once we use this script more.
 
>     * [ ]  have you considered mounting in the containers `~/.bashrc`? It oculd be useful to have the history of commands so we don't have to repeatedly type them.

Good question. You probably refer to the `.bash_history` file, and I've in fact considered it. Currently, the state of the Dangerzone environment is just the Podman containers (`/home/user/.local/share/containers`), but that wasn't always the case. I used to mount the whole user's home (`/home/user`) as a directory, meaning that all user state would persist across executions (containers, bash history, Dangerzone config, you name it).

I dropped it because I need to think a bit if past state can affect subsequent executions. Most probably it won't, but it doesn't hurt us delaying this a bit. Are you ok with that?

> 
>     * [ ]  The build `.deb` step was much slower for me. It took 27mins. Is this slow performance to be expected?

No, it shouldn't. I haven't experience such slow build times myself. We may need to debug this.

>     * [ ]  In the container there is no PDF viewer, which leads to
>       ![no-pdf-viewer](https://user-images.githubusercontent.com/47065258/209135525-c9663944-58e1-48aa-bef8-b1c554b2b824.png)

I personally install `mupdf`, which is a light-weight PDF viewer, when running the QA steps. Do you have a PDF viewer that does not bring much dependencies?

>     * when an 'auto' is running and you press any key when the command is running, those keys will get pressed after the command runs. So if you press enter while waiting to see if something happens, then it presses enter on the next commands -- but because enter skips the task, you advance to the next ones. I don't know how to solve this problem easily, but changing the auto from `enter` to something else less clickable can help.

Good catch. I've mentioned in the corresponding discussion (https://github.com/freedomofpress/dangerzone/pull/289#discussion_r1060568969) that we can consume the process' stdin before prompting the user, which would solve these cases. We can discuss more there.

> ### Some situations where 'auto' failed:
> 
>     * I've had many 'auto' commands related to lack of space. But that's to be expected. After increasing the size the went away. Because this may be so common, should we add a check for them?

Hm, this gun shoots in two directions. Imagine the case where a step can run without extra space, but the system is low in resources. In that case, what will this check do? Cancel a reasonable execution? Throw a warning that will be lost in the log messages. I'm not sure how we can make this work, but if you have a suggestion, shoot it. Disclaimer, most of the issues I've encountered where related to low disk space as well, so I get your point.

> 
>     * Failure in auto step "Create Dangerzone QA environment" (fedora 36)
> 
> I was unable to find why this was failed, but after repeating the step, it worked.

Interesting, I haven't seen this before. Is this reproducible? If yes, we can look into it. Else, I'd attribute it to something transient (maybe low disk space).

deeplow commented:
> > ```
> > * [ ]  have you considered mounting in the containers `~/.bashrc`? It oculd be useful to have the history of commands so we don't have to repeatedly type them.
> > ```
> 
> 
> Good question. You probably refer to the `.bash_history` file, and I've in fact considered it. Currently, the state of the Dangerzone environment is just the Podman containers (`/home/user/.local/share/containers`), but that wasn't always the case. I used to mount the whole user's home (`/home/user`) as a directory, meaning that all user state would persist across executions (containers, bash history, Dangerzone config, you name it).
> 
> I dropped it because I need to think a bit if past state can affect subsequent executions. Most probably it won't, but it doesn't hurt us delaying this a bit. Are you ok with that?

Yes, I meant `.bash_history`. I get the point of keeping the state. But I wouldn't consider keeping past commands in history as state that will be of terrible consequence. If anything it will speed up testing. It's a quality-of-life improvement that I think would make a difference in practice. So I'd insist for us to keep it.

> I personally install mupdf, which is a light-weight PDF viewer, when running the QA steps. Do you have a PDF viewer that does not bring much dependencies?

Nope. It can be that one or anything else you prefer.



> Hm, this gun shoots in two directions. Imagine the case where a step can run without extra space, but the system is low in resources. In that case, what will this check do? Cancel a reasonable execution? Throw a warning that will be lost in the log messages. I'm not sure how we can make this work, but if you have a suggestion, shoot it. Disclaimer, most of the issues I've encountered where related to low disk space as well, so I get your point.

Just a warning would be my suggestion + the use of colors (bold, yellow). But we can tackle this in the future when we consider colors in the script.


> Interesting, I haven't seen this before. Is this reproducible? If yes, we can look into it. Else, I'd attribute it to something transient (maybe low disk space).

Not reproducible. Let's just discard this issue.

apyrgio commented:
Hi @deeplow. I think I've addressed all the review comments that we talked about, and I fixed some issues tangent to this code. Feel free to give me your opinion on the changes.

deeplow commented on dev_scripts/env.py:
```diff
@@ -0,0 +1,527 @@
+        # Mount the source and the state of the distro into the container
+        dz_src = git_root()
+        dist_state = distro_state(self.distro, self.version)
+        run_cmd += [
+            "-v",
+            f"{dz_src}:/home/user/dangerzone",
+            "-v",
+            f"{dist_state}/containers:/home/user/.local/share/containers",
+            "-v",
+            f"{dist_state}/.bash_history:/home/user/.bash_history",
+        ]
```
Ah, you interpreted keeping `.bash_history` from the same container. I actually meant mounting the history from the user's system. This way it would be shared in all containers. But I guess this is fine as well and may lead to less problems (e.g. when devs use another shell).

apyrgio replied:
Another problem with that would be that Fedora envs would see APT commands, and your host would see Dangerzone commands. Probably best to keep those separate.

deeplow reviewed (approved):
Only minor comments / questions. No need to address them as they are minor.

Approved in general

apyrgio commented on dev_scripts/qa.py:
```diff
@@ -0,0 +1,885 @@
+
+        def decorator(func):
+            def inner(self, *args, **kwargs):
+                # If the reference is actually defined in subclasses, grab it from
+                # `self`.
+                if isinstance(ref, str):
+                    _ref = getattr(self, ref)
+                else:
+                    _ref = ref
```
Sigh, I'm not proud of this code. An example where the problem manifests is here:

https://github.com/freedomofpress/dangerzone/blob/f0c3c82f25e77b3e59a30a776e926a5a840fbcc0/dev_scripts/qa.py#L699-L700

If you look carefully, I use `ref="REF_BUILD"`, instead of `ref=REF_BUILD`. The reason is that during class creation, `REF_BUILD = None`. To accomodate for the `REF_BUILD` that the subclasses later define, I had to resort to a dynamic approach :disappointed: .

apyrgio commented:
I've rebased the branch, and I've removed the Fedora 35 support, so now our build completes successfully.
